bot/slack_im_notifier.go

```diff
@@ -3,26 +3,52 @@ package bot
 import (
 	"fmt"
 	"log"
+	"sync"
+	"time"
 
-	"github.com/andrewslotin/michael/deploy"
-	"github.com/andrewslotin/michael/slack"
+	"github.com/adjust/michaelbot/deploy"
+	"github.com/adjust/michaelbot/slack"
 )
 
 type SlackIMNotifier struct {
-	im    *slack.InstantMessenger
-	users *slack.TeamDirectory
+	im             *slack.InstantMessenger
+	users          *slack.TeamDirectory
+	warningTimeout time.Duration
+	timer          *time.Timer
+	mutex          sync.Mutex
 }
 
-func NewSlackIMNotifier(api *slack.WebAPI) *SlackIMNotifier {
+func NewSlackIMNotifier(api *slack.WebAPI, warningTimeout time.Duration) *SlackIMNotifier {
 	return &SlackIMNotifier{
-		im:    slack.NewInstantMessenger(api),
-		users: slack.NewTeamDirectory(api),
+		im:             slack.NewInstantMessenger(api),
+		users:          slack.NewTeamDirectory(api),
+		warningTimeout: warningTimeout,
 	}
 }
 
-func (notifier *SlackIMNotifier) DeployStarted(_ string, _ deploy.Deploy) {}
+func (notifier *SlackIMNotifier) DeployStarted(_ string, d deploy.Deploy) {
+	notifier.mutex.Lock()
+	defer notifier.mutex.Unlock()
+
+	notifier.timer = time.AfterFunc(notifier.warningTimeout, func() {
```

Comment:
Don't you also need to stop the previous timer on `DeployStart`? I haven't checked the whole flow, but will this work correctly if a user started a deploy and then started another deploy? Like

```
/deploy <github/pull/1>
...
/deploy <github/pull/2>
...
/deploy done
```

Who will remove the goroutine from the first deploy?

Reply:
good point 👍